### PR TITLE
enable linting and formatting for JS code

### PR DESCRIPTION
**Pull request by shiftkey**
I'm terrified of breaking stuff in #1151 so I wanted to step back and look at getting some checks in place before I start making radical changes to the JS.

The first thing that came to mind was linting the code. Why?

 - there's a healthy ecosystem of tools out there to lint your JS
 - this can be applied gradually, changing the JS only when necessary
 - this can be integrated into CI, and fail the build when an issue is introduced

This PR does a few things:

 - add in a `package.json` to enable tooling and scripting
 - configure `eslint` to run on our source and test JS files
 - mute a whole bunch of stuff for the moment
 - address a subset of easy changes
 - wire up CI so that can be automated when submitting PRs

TODO:

 - [x] cleanup some more `eslint` rules so that the list of exclusions is a sane subset
 - [x] documentation for working with these tools
 - [x] feedback from others?

### CODE REVIEW

**Comment by shiftkey** [comment]
After sleeping on this some more, I've settled on some specific goals on top of addressing the two bugs I called out in #1151:

 - have a consistent coding style applied to the JS, which is enforced via tooling and CI
 - have automated tests for features, which can be run locally via the command line and are also enforced via CI

This PR helps with the first case, but the second part requires some extra work. I have an experimental branch lying around that has these goals:

 - use `jest` as the test runner
 - port existing `jasmine` tests to be consumed in `jest`
 - get the tests running

Because of the use of [`IIFE`](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) functions that bind to globals in areas like the project service, this requires reworking some of the core parts to be able to be `export`-ed and `require`-d. 

I still want this site to be able to be run without needing a bundler or manually committing changes to the underlying libraries, but I just wanted to write up where my head is at after coming back to the codebase after some significant time away.

I'd love to hear from the maintainers about going down this path before I go further - cc @dahlbyk @ritwik12 @jrusbatch - and also the peeps who have been working inside the main JS files over the past year or so - cc @wolffles @juliankrieger @Zarad1993 @matt011 @Motardo @richa208 @djm158 @evanandrewrose. 

A special shout out to @Zarad1993 for adding to test suite - I'd kind of forgotten they existed!

**Comment by Zarad1993** [comment]
@shiftkey - I highly recommend the movement of this PR, it will definitely help keep the project cleaner and readable for future contributors. Let me know if there is anything I can help with moving it forward.

**Comment by ritwik12** [comment]
@shiftkey I guess both the points are very much needed. As being a maintainer I have faced some hard time reviewing and merging new feature PRs and a few times our site went down due to those. Few features works well locally according to contributors but fails to do so upstream. I think having a  consistent coding style applied to the JS and automated tests for features are pretty much required.

**Comment by jrusbatch** [comment]
@shiftkey After also coming back to this codebase after some significant time away, I think these changes (and going forward with automated tests than can be run locally and via CI) are a great idea.

**Review by jrusbatch — APPROVED** [review]

File: docs/setup.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+You should also ensure the code satisfies the rules defined in the `eslintrc.json`
+config file:
+
+```shellsession
+$ yarn lint
+```
+
+These rules will be checked as part of every pull request into the project, but
+please ensure you run these commands locally to catchi any issues before a
```
Comment (jrusbatch) [inline docs/setup.md]: Typo here: `catchi`

Comment (shiftkey) [reply docs/setup.md]: For future reference: https://twitter.com/github/status/1053360435074949120

**Pull request merged**